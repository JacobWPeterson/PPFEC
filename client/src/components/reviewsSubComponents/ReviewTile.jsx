/* eslint-disable camelcase */
import React from 'react';
import styled from 'styled-components';
import RatingStars from './RatingStars.jsx';

const TileContainer = styled.div`
  margin: 0 auto;
  padding: 10px 30px;
  border-bottom: 1px solid black;
`;

const monthNames = ['January', 'February', 'March', 'April', 'May', 'June',
  'July', 'August', 'September', 'October', 'November', 'December'];

const StarDateWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: baseline;
`;

const ResponseWrapper = styled.div`
  margin: 0 auto;
  backgound-color: lighblue;
`;

const ReviewThumbsWrapper = styled.img`
  display: inline-flex;
  justify-content: space-around;
  align-items: flex-star;
  border: 1px solid #ddd;
  padding: 5px;
  width: 80px;

  & :hover {
    box-shadow: 0 0 2px 1px rgba(0, 140, 186, 0.5);
  }
`;

class ReviewTile extends React.Component {
  constructor(props) {
    super(props);
    this.state = {

    };
  }

  render() {
    const { review } = this.props;
    const {
      date, rating, reviewer_name, summary, body, helpfulness,
    } = review;
    // Formatting date
    const reviewDate = new Date(date);
    const month = monthNames[reviewDate.getUTCMonth()];
    const day = reviewDate.getUTCDate();
    const year = reviewDate.getUTCFullYear();

    // calculating the percentage for stars
    const percentage = (rating / 5) * 100;

    // conditional rendering for recommend label
    let recommendLabel;
    let response;
    if (review.recommend) {
      recommendLabel = <div>âœ“ I recommend this product</div>;
    }
    // conditional rendering for response
    if (review.response) {
      response = (
        <ResponseWrapper>
          <h4>Response from seller</h4>
          <p>{review.response}</p>
        </ResponseWrapper>
      );
    }
    // conditional rendering for photos
    let photos;
    if (review.photos.length > 0) {
      photos = (
        <div>
          {review.photos.map((photo, index) => (
            <ReviewThumbsWrapper key={photo.id} src={photo.url} alt={`${index}reviewPhoto`} />
          ))}
        </div>
      );
    }

    return (
      <TileContainer>
        <StarDateWrapper>
          <RatingStars percent={percentage} />
          <span>
            { reviewer_name }
            ,
            {' '}
            {month}
            {' '}
            {day}
            ,
            {' '}
            {year}
          </span>
        </StarDateWrapper>
        <h3>{summary}</h3>
        <p>{body}</p>
        {photos}
        {recommendLabel}
        {response}
        <div>
          <span>
            Helpful?
            <a href="">Yes</a>
            (
            {helpfulness}
            )
          </span>
          {' '}
          |
          {' '}
          <span><a href="">Report</a></span>
        </div>
      </TileContainer>
    );
  }
}

export default ReviewTile;
