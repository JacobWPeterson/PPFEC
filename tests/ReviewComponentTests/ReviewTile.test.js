/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { shallow, mount, render } from 'enzyme';
import waitUntil from "async-wait-until";
import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import _ from "lodash";
import ReviewTile from '../../client/src/components/reviewsSubComponents/ReviewTile.jsx';

describe('ReviewTile Component', () => {
  jest.mock('axios', () => ({
    __esModule: true,
    default: jest.fn(),
  }));

  const props = {
    review: {
      review_id: 147659,
      rating: 2,
      summary: 'Tenetur quo blanditiis necessitatibus eveniet quo ipsa quia alias voluptas.',
      recommend: true,
      response: null,
      body: 'Non amet dolorem ut tempora perferendis sint ut adipisci suscipit. Sint ea itaque debitis quos enim odit. Ad perferendis nihil quisquam est. Laboriosam fuga ea amet repellat. Blanditiis aspernatur nostrum aspernatur quas necessitatibus.',
      date: '2021-01-26T00:00:00.000Z',
      reviewer_name: 'Shyanne6',
      helpfulness: 12,
      photos: [{ id: 187178, url: 'https://images.unsplash.com/photo-1469617833234-8a6843da14d0?ixlib=rb-1.2.1&auto=format&fit=crop&w=2764&q=80' }, { id: 187157, url: 'https://images.unsplash.com/uploads/1412198532414025532c0/6a31309c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1600&q=80' }],
    },
  };
  let ReviewTileComponent;
  beforeEach(() => {
    ReviewTileComponent = mount(<ReviewTile {...props} debug />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Should render correctly', () => {
    expect(ReviewTileComponent).toMatchSnapshot();
  });

  it('Should render all the photos if there are photos from the data', () => {
    const photoNumber = ReviewTileComponent.find('img').length;
    expect(photoNumber).toEqual(props.review.photos.length);
  });

  // it('Should increment the helpfulness number when clicking "Yes" label', () => {
  //   jest.spyOn(axios, 'dafault').mockResolvedValue({});
  //   const spyOnYesClick = jest.spyOn(ReviewTileComponent.instance(), 'handleClickYes');
  //   // const mockPreventDefault = jest.fn();
  //   // const mockPersist = jest.fn();
  //   // const mockEvent = {
  //   //   preventDefault: mockPreventDefault,
  //   //   persist: mockPersist,
  //   // };
  //   const yesClick = ReviewTileComponent.find('.not-click').at(0);
  //   ReviewTileComponent.instance().handleClickYes();
  //   yesClick.simulate('click');
  //   yesClick.props('onClick');
  //   expect(spyOnYesClick).toBeCalled();
  //   process.nextTick(() => {
  //     expect(ReviewTileComponent.state('yesNum')).toBe(props.review.helpfulness + 1);
  //   });
  // });
});
