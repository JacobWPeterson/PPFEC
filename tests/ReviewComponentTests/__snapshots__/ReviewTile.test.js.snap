// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewTile Component Should render correctly 1`] = `
<ReviewTile
  debug={true}
  review={
    Object {
      "body": "Non amet dolorem ut tempora perferendis sint ut adipisci suscipit. Sint ea itaque debitis quos enim odit. Ad perferendis nihil quisquam est. Laboriosam fuga ea amet repellat. Blanditiis aspernatur nostrum aspernatur quas necessitatibus.",
      "date": "2021-01-26T00:00:00.000Z",
      "helpfulness": 12,
      "photos": Array [
        Object {
          "id": 187178,
          "url": "https://images.unsplash.com/photo-1469617833234-8a6843da14d0?ixlib=rb-1.2.1&auto=format&fit=crop&w=2764&q=80",
        },
        Object {
          "id": 187157,
          "url": "https://images.unsplash.com/uploads/1412198532414025532c0/6a31309c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1600&q=80",
        },
      ],
      "rating": 2,
      "recommend": true,
      "response": null,
      "review_id": 147659,
      "reviewer_name": "Shyanne6",
      "summary": "Tenetur quo blanditiis necessitatibus eveniet quo ipsa quia alias voluptas.",
    }
  }
>
  <ReviewTile__TileContainer
    reportClick={false}
  >
    <div
      className="ReviewTile__TileContainer-sc-1vbuo4p-0 izQjHl"
    >
      <ReviewTile__StarDateWrapper>
        <div
          className="ReviewTile__StarDateWrapper-sc-1vbuo4p-1 hAYJuv"
        >
          <RatingStars
            rating={2}
          >
            <div>
              <RatingStars__StarsContainer>
                <div
                  className="RatingStars__StarsContainer-sc-1djn8fm-0 cYnGWG"
                >
                  <RatingStars__StarsContainer
                    className="top"
                    rating={2}
                  >
                    <div
                      className="RatingStars__StarsContainer-sc-1djn8fm-0 btLlnX top"
                    >
                      <span>
                        ★
                      </span>
                      <span>
                        ★
                      </span>
                      <span>
                        ★
                      </span>
                      <span>
                        ★
                      </span>
                      <span>
                        ★
                      </span>
                    </div>
                  </RatingStars__StarsContainer>
                  <RatingStars__StarsContainer
                    className="bottom"
                  >
                    <div
                      className="RatingStars__StarsContainer-sc-1djn8fm-0 cYnGWG bottom"
                    >
                      <span>
                        ★
                      </span>
                      <span>
                        ★
                      </span>
                      <span>
                        ★
                      </span>
                      <span>
                        ★
                      </span>
                      <span>
                        ★
                      </span>
                    </div>
                  </RatingStars__StarsContainer>
                </div>
              </RatingStars__StarsContainer>
            </div>
          </RatingStars>
          <span>
            Shyanne6
            ,
             
            January
             
            26
            ,
             
            2021
          </span>
        </div>
      </ReviewTile__StarDateWrapper>
      <ReviewTile__StyledSummary>
        <div
          className="ReviewTile__StyledSummary-sc-1vbuo4p-6 hKqFGZ"
        >
          Tenetur quo blanditiis necessitatibus eveniet quo ipsa quia alias voluptas.
        </div>
      </ReviewTile__StyledSummary>
      <ReviewTile__StyledBody>
        <p
          className="ReviewTile__StyledBody-sc-1vbuo4p-7 hxWBm"
        >
          Non amet dolorem ut tempora perferendis sint ut adipisci suscipit. Sint ea itaque debitis quos enim odit. Ad perferendis nihil quisquam est. Laboriosam fuga ea amet repellat. Blanditiis aspernatur nostrum aspernatur quas necessitatibus.
        </p>
      </ReviewTile__StyledBody>
      <div>
        <ReviewTile__ReviewThumbsWrapper
          alt="0reviewPhoto"
          key="187178"
          src="https://images.unsplash.com/photo-1469617833234-8a6843da14d0?ixlib=rb-1.2.1&auto=format&fit=crop&w=2764&q=80"
        >
          <img
            alt="0reviewPhoto"
            className="ReviewTile__ReviewThumbsWrapper-sc-1vbuo4p-3 bchCvd"
            src="https://images.unsplash.com/photo-1469617833234-8a6843da14d0?ixlib=rb-1.2.1&auto=format&fit=crop&w=2764&q=80"
          />
        </ReviewTile__ReviewThumbsWrapper>
        <ReviewTile__ReviewThumbsWrapper
          alt="1reviewPhoto"
          key="187157"
          src="https://images.unsplash.com/uploads/1412198532414025532c0/6a31309c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1600&q=80"
        >
          <img
            alt="1reviewPhoto"
            className="ReviewTile__ReviewThumbsWrapper-sc-1vbuo4p-3 bchCvd"
            src="https://images.unsplash.com/uploads/1412198532414025532c0/6a31309c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1600&q=80"
          />
        </ReviewTile__ReviewThumbsWrapper>
      </div>
      <ReviewTile__StyledRecommend>
        <div
          className="ReviewTile__StyledRecommend-sc-1vbuo4p-8 hHkGbK"
        >
          ✓ I recommend this product
        </div>
      </ReviewTile__StyledRecommend>
      <ReviewTile__HelpfulnessWrapper>
        <div
          className="ReviewTile__HelpfulnessWrapper-sc-1vbuo4p-4 kfimVf"
        >
          Helpful?
          <ReviewTile__ClickTag
            className="not-click"
            onClick={[Function]}
          >
            <div
              className="ReviewTile__ClickTag-sc-1vbuo4p-5 jSpjYR not-click"
              onClick={[Function]}
            >
              Yes
               
              (
              12
              )
            </div>
          </ReviewTile__ClickTag>
          <div>
              |  
          </div>
          <ReviewTile__ClickTag
            className="not-click"
            onClick={[Function]}
          >
            <div
              className="ReviewTile__ClickTag-sc-1vbuo4p-5 jSpjYR not-click"
              onClick={[Function]}
            >
              Report
            </div>
          </ReviewTile__ClickTag>
        </div>
      </ReviewTile__HelpfulnessWrapper>
    </div>
  </ReviewTile__TileContainer>
</ReviewTile>
`;
